argoverse/evaluation/eval_detection.py

```diff
@@ -0,0 +1,412 @@
+        # The following line grabs the max affinity for each detection to a ground truth label.
+        affinities = np.take_along_axis(affinity_matrix.T, gt_matches, axis=0).squeeze(0)
+
+        # Find the indices of the "first" detection assigned to each GT.
+        unique_gt_matches, unique_dt_matches = np.unique(gt_matches, return_index=True)
+        for i, thresh in enumerate(self.detection_cfg.affinity_threshs):
+
+            # `tp_mask` may need to be defined differently with other affinities
+            tp_mask = affinities[unique_dt_matches] > -thresh
+            metrics[unique_dt_matches, i] = tp_mask
+
+            # Only compute true positive error when `thresh` is equal to the tp threshold.
+            is_tp_thresh = thresh == self.detection_cfg.tp_thresh
+            # Ensure that there are true positives of the respective class in the frame.
+            has_true_positives = np.count_nonzero(tp_mask) > 0
+
+            if is_tp_thresh and has_true_positives:
+                dt_tp_indices = unique_dt_matches[tp_mask]
+                gt_tp_indices = unique_gt_matches[tp_mask]
+
+                # Form DataFrame of shape (N, D) where D is the number of attributes in `ObjectLabelRecord`.
+                dt_df = pd.DataFrame([dt.__dict__ for dt in dts[dt_tp_indices]])
+                gt_df = pd.DataFrame([gt.__dict__ for gt in gts[gt_tp_indices]])
+
+                trans_error = dist_fn(dt_df, gt_df, DistFnType.TRANSLATION)
+                scale_error = dist_fn(dt_df, gt_df, DistFnType.SCALE)
+                orient_error = dist_fn(dt_df, gt_df, DistFnType.ORIENTATION)
+
+                metrics[dt_tp_indices, n_threshs : n_threshs + N_TP_ERRORS] = np.vstack(
+                    (trans_error, scale_error, orient_error)
+                ).T
+        return metrics
+
+    def summarize(
+        self, data: DefaultDict[str, np.ndarray], cls_to_ninst: DefaultDict[str, int]
+    ) -> DefaultDict[str, List]:
+        """Calculate and print the detection metrics.
+
+        Args:
+            data: The aggregated data used for summarization.
+            cls_to_ninst: Map of classes to number of instances.
+
+        Returns:
+            summary: The summary statistics.
+        """
+        summary: DefaultDict[str, List] = defaultdict(list)
+        recalls_interp = np.linspace(0, 1, self.detection_cfg.n_rec_samples)
+        num_ths = len(self.detection_cfg.affinity_threshs)
+        if not self.figures_fpath.is_dir():
+            self.figures_fpath.mkdir(parents=True, exist_ok=True)
+
+        for cls_name, cls_stats in data.items():
+            ninst = cls_to_ninst[cls_name]
+            ranks = cls_stats[:, -1].argsort()[::-1]
+            cls_stats = cls_stats[ranks]
+
+            for i, _ in enumerate(self.detection_cfg.affinity_threshs):
+                tp = cls_stats[:, i].astype(bool)
+
+                cumulative_tp = np.cumsum(tp, dtype=np.int)
```

Comment:
Might be a bit safer to make a function dedicated just to AP so that you can write a unit test for it:

```python
ap_th, recs_interp, precs_interp = calc_ap(
     gts_sorted_by_conf=cls_stats[:, i].astype(bool),
     n_rec_samples=self.detection_cfg.n_rec_samples
)
```

defined as:
```python
def calc_ap(gts_sorted_by_conf, n_rec_samples: int) -> Tuple[float, np.ndarray, np.ndarray]
     """ Compute precision and recall, interpolated over n fixed recall points """
       tp = cls_stats[:, i].astype(bool)

       cumulative_tp = np.cumsum(tp, dtype=np.int)
       cumulative_fp = np.cumsum(~tp, dtype=np.int)
       cumulative_fn = ninst - cumulative_tp

       precisions = cumulative_tp / (cumulative_tp + cumulative_fp + np.finfo(float).eps)
       recalls = cumulative_tp / (cumulative_tp + cumulative_fn)
       precisions = interp(precisions)
       precisions_interp = np.interp(recalls_interp, recalls, precisions, right=0)
       ap_th = precisions_interp.mean()
      return ap_th, precisions_interp, recalls_interp
```

Reply:
i think the function is already too long as it is (63 lines, from 312-375)